### PR TITLE
PS-5223 : Can't upgrade with encrypted undo tablespaces from 5.7.24 to

### PR DESCRIPTION
**Pull request by robgolebiowski**
8.0.13

This is a fix for upgrade from 5.7 encrypted undo tablespaces to PS
8.0.15. The fix is to accept encrypted undo tablespace header with
ENCRYPTION_KEY_MAGIC_V2 (instead of only ENCRYPTION_KEY_MAGIC_V3),
since PS 5.7 encrypted undo tablespaces are encrypted with
ENCRYPTION_KEY_MAGIC_V2.

Testcase percona_dd_upgrade_undo_encrypted was addded. It tests
upgrade with 5.7 encrypted undo tablespaces with:
1) correct keyring and later upgrade
2) empty keyring
3) corrupted undo encryption keys

### CODE REVIEW

**Comment by robgolebiowski** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-trunk/9/

File: storage/innobase/srv/srv0start.cc [hunk 1]
```diff
@@ -688,6 +688,9 @@ static dberr_t srv_undo_tablespace_read_encryption(pfs_os_file_t fh,
 
   /* Return if the encryption metadata is empty. */
   if (memcmp(first_page + offset, ENCRYPTION_KEY_MAGIC_V3,
+             ENCRYPTION_MAGIC_SIZE) != 0 &&
+      /* PS 5.7 undo encryption */
+      memcmp(first_page + offset, ENCRYPTION_KEY_MAGIC_V2,
```
Comment (satya-bodapati) [inline storage/innobase/srv/srv0start.cc]: Shouldn't the exemption be given only when srv_is_upgrade_mode is true?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+--echo # Set path for undo* files.
+let $MYSQLD_UNDO_DATADIR = $MYSQL_TMP_DIR/undo57_encrypted/innodb_undo_data_dir;
+
+--echo # Set path for ibdata* files.
+let $MYSQLD_HOME_DATA_DIR = $MYSQL_TMP_DIR/undo57_encrypted/innodb_data_home_dir;
+
+--copy_file $MYSQLTEST_VARDIR/std_data/undo57_encrypted.zip $MYSQL_TMP_DIR/undo57_encrypted.zip
+
+--echo # Check that the file exists in the working folder.
+--file_exists $MYSQL_TMP_DIR/undo57_encrypted.zip
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_dd_upgrade_undo_encrypted.test]: Redundant, let either copy_file, either exec unzip fail

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+--exec $MYSQL_UPGRADE --skip-verbose --force 2>&1
+--source include/mysql_upgrade_cleanup.inc
+
+
+SHOW CREATE TABLE test.tab1;
+SELECT * FROM tab1;
+
+--source include/shutdown_mysqld.inc
+
+--echo # Now lets test what happens when encryption key is missing and we try to start 8.0 server on 5.7 directory
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_dd_upgrade_undo_encrypted.test]:
```suggestion
--echo # Now let's test what happens when encryption key is missing and we try to start 8.0 server on 5.7 directory
```

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+--echo # Set path for ibdata* files.
+let $MYSQLD_HOME_DATA_DIR = $MYSQL_TMP_DIR/undo57_encrypted/innodb_data_home_dir;
+
+--copy_file $MYSQLTEST_VARDIR/std_data/undo57_encrypted.zip $MYSQL_TMP_DIR/undo57_encrypted.zip
+
+--echo # Check that the file exists in the working folder.
+--file_exists $MYSQL_TMP_DIR/undo57_encrypted.zip
+
+--echo # Unzip the zip file.
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_dd_upgrade_undo_encrypted.test]: Too obvious comment
```suggestion
```

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+SELECT * FROM tab1;
+
+--source include/shutdown_mysqld.inc
+
+--echo # Now lets test what happens when encryption key is missing and we try to start 8.0 server on 5.7 directory
+--echo # with encrypted undo tablespaces
+--echo # Recreate the directories
+--force-rmdir $MYSQL_TMP_DIR/undo57_encrypted
+--echo # Check that the file exists in the working folder.
+--file_exists $MYSQL_TMP_DIR/undo57_encrypted.zip
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_dd_upgrade_undo_encrypted.test]: Remove

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+--echo # Unzip the zip file.
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_dd_upgrade_undo_encrypted.test]:
```suggestion
```

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/innobase/srv/srv0start.cc (hunk 1 above)
Comment (robgolebiowski) [reply storage/innobase/srv/srv0start.cc]: DD where the upgrade mode is detected is initialized after the undo tablespaces are opened thus - no.

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/t/percona_dd_upgrade_undo_encrypted.test]: I cannot remove the --force-rmdir part as there are some leftovers and server refuses to start:
2019-04-05T16:27:38.197914Z 1 [ERROR] [MY-012209] [InnoDB] Multiple files found for the same tablespace ID:
2019-04-05T16:27:38.197938Z 1 [ERROR] [MY-012202] [InnoDB] Tablespace ID: 15 = ['mysql/innodb_table_stats.ibd', 'mysql/innodb_table_stats_backup57.ibd']

2019-04-05T16:27:38.197955Z 1 [ERROR] [MY-012202] [InnoDB] Tablespace ID: 16 = ['mysql/innodb_index_stats.ibd', 'mysql/innodb_index_stats_backup57.ibd']

File: storage/innobase/srv/srv0start.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/srv/srv0start.cc]: This answer should be "yes". I am not convinced with the answer and I tried myself.  We can rely on the variable "srv_is_upgrade_mode".

(gdb) p srv_is_upgrade_mode
$4 = true

(gdb) p space->name
$5 = 0x7fffe2c63618 "innodb_undo001"

(gdb) p space->files
$6 = std::vector of length 1, capacity 1 = {{space = 0x7fffe2c63358, name = 0x7fffe2c63678 "./undo001", is_open = false, handle = {m_psi = 0x0,
      m_file = 0}, sync_event = 0x7fffe2c636a8, is_raw_disk = false, size = 640, flush_size = 640, init_size = 640, max_size = 4294967294,
    n_pending = 0, n_pending_flushes = 0, in_use = 0, modification_counter = 0, flush_counter = 0, LRU = {prev = 0x0, next = 0x0}, punch_hole = true,
    block_size = 4096, atomic_write = false, magic_n = 89389}}

#0  srv_undo_tablespace_read_encryption (fh=..., space=0x7fffe2c63358) at /home/satya/WORK/PS-8.0-5508/storage/innobase/srv/srv0start.cc:659
#1  0x0000555559c3b6c5 in srv_undo_tablespace_open (undo_space=...) at /home/satya/WORK/PS-8.0-5508/storage/innobase/srv/srv0start.cc:949
#2  0x0000555559c3ba14 in srv_undo_tablespace_open_by_id (space_id=1) at /home/satya/WORK/PS-8.0-5508/storage/innobase/srv/srv0start.cc:997
#3  0x0000555559c3c03b in srv_undo_tablespaces_open () at /home/satya/WORK/PS-8.0-5508/storage/innobase/srv/srv0start.cc:1090
#4  0x0000555559c3d4d1 in srv_undo_tablespaces_init (create_new_db=false) at /home/satya/WORK/PS-8.0-5508/storage/innobase/srv/srv0start.cc:1506
#5  0x0000555559c40ef2 in srv_start (create_new_db=false, scan_directories="./;./;./")
    at /home/satya/WORK/PS-8.0-5508/storage/innobase/srv/srv0start.cc:2755
#6  0x0000555559988a5c in innobase_init_files (dict_init_mode=DICT_INIT_UPGRADE_57_FILES, tablespaces=0x7fffe865f930)
    at /home/satya/WORK/PS-8.0-5508/storage/innobase/handler/ha_innodb.cc:5356
#7  0x000055555999a0f5 in innobase_ddse_dict_init (dict_init_mode=DICT_INIT_UPGRADE_57_FILES, version=80014, tables=0x7fffe865f910, 
    tablespaces=0x7fffe865f930) at /home/satya/WORK/PS-8.0-5508/storage/innobase/handler/ha_innodb.cc:12586
#8  0x00005555594ab2e8 in dd::bootstrap::DDSE_dict_init (thd=0x55555c473f50, dict_init_mode=DICT_INIT_UPGRADE_57_FILES, version=80014)
    at /home/satya/WORK/PS-8.0-5508/sql/dd/impl/bootstrapper.cc:2320
#9  0x00005555596da55d in dd::upgrade_57::do_pre_checks_and_initialize_dd (thd=0x55555c473f50)
    at /home/satya/WORK/PS-8.0-5508/sql/dd/upgrade/upgrade.cc:1086
#10 0x000055555874dff2 in bootstrap::handle_bootstrap (arg=0x55555c473f50) at /home/satya/WORK/PS-8.0-5508/sql/bootstrap.cc:347
#11 0x0000555559892320 in pfs_spawn_thread (arg=0x55555c371320) at /home/satya/WORK/PS-8.0-5508/storage/perfschema/pfs.cc:2836
#12 0x00007ffff79b66db in start_thread (arg=0x7fffe8660700) at pthread_create.c:463
#13 0x00007ffff5e9388f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95



ls data/
auto.cnf      ib_buffer_pool  ib_logfile0  mtr    mysql_dd_upgrade_info  sys   undo001  undo003
binlog.index  ibdata1         ib_logfile1  mysql  performance_schema     test  undo002  xb_doublewrite

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test (hunk 5 above)
Comment (laurynas-biveinis) [reply mysql-test/t/percona_dd_upgrade_undo_encrypted.test]: No remove not the force-rmdir but file_exists line

File: storage/innobase/srv/srv0start.cc (hunk 1 above)
Comment (robgolebiowski) [reply storage/innobase/srv/srv0start.cc]: sorry, I have messed up the condition when I was testing this. You are right.

File: mysql-test/t/percona_dd_upgrade_undo_encrypted.test (hunk 5 above)
Comment (robgolebiowski) [reply mysql-test/t/percona_dd_upgrade_undo_encrypted.test]: ack, this was done

File: storage/innobase/srv/srv0start.cc (hunk 1 above)
Comment (satya-bodapati) [reply storage/innobase/srv/srv0start.cc]: Thanks, Will wait for the updated patch

**Comment by robgolebiowski** [comment]
https://ps80.cd.percona.com/job/percona-server-8.0-trunk/11/

**Comment by robgolebiowski** [comment]
@satya-bodapati @laurynas-biveinis @dutow ready for review

**Review by laurynas-biveinis — APPROVED** [review]

**Review by satya-bodapati — APPROVED** [review]

**Review by dutow — APPROVED** [review]

**Pull request merged**